src/imap/imap-fetch.c

```diff
@@ -916,6 +916,66 @@ static bool imap_fetch_savedate_init(struct imap_fetch_init_context *ctx)
 	return TRUE;
 }
 
+static int fetch_emailid(struct imap_fetch_context *ctx, struct mail *mail,
+		      void *context ATTR_UNUSED)
+{
+	const char *value;
+
+	if (mail_get_special(mail, MAIL_FETCH_GUID, &value) < 0)
+		return -1;
+
+	str_printfa(ctx->state.cur_str, "EMAILID (%s) ", value);
+	return 1;
+}
+
+static bool fetch_emailid_init(struct imap_fetch_init_context *ctx)
+{
+	ctx->fetch_ctx->fetch_data |= MAIL_FETCH_GUID;
+	imap_fetch_add_handler(ctx, IMAP_FETCH_HANDLER_FLAG_BUFFERED,
+			       "", fetch_emailid, NULL);
+	return TRUE;
+}
+
+static int fetch_threadid(struct imap_fetch_context *ctx, struct mail *mail,
+		      void *context ATTR_UNUSED)
+{
+	const char *value;
+
+	if (mail_get_special(mail, MAIL_FETCH_GUID, &value) < 0)
+		return -1;
+
+	str_printfa(ctx->state.cur_str, "THREADID (%s) ", value);
```

Comment:
I think this should be NIL until we actually support it.

Reply:
ok, any ideas how could by THREADID implemented?

Reply:
And as for answer to this: With lots of pain and effort. My thinking has been to create one or two new index files where you can get msgid <-> threadid mappings. I've a more detailed (but not finished) plan written out (DOP-768 in our internal Jira).